server/src/service-layer/controllers/PaymentController.ts

```diff
@@ -176,4 +184,182 @@ export class PaymentController extends Controller {
+      return {
+        error: "Invalid date range, booking must be a maximum of 10 days. "
+      }
+    }
+    // Create new Stripe checkout session
+    const stripeService = new StripeService()
+    const userDataService = new UserDataService()
+
+    const userData = await userDataService.getUserData(uid)
+    const { newUser, stripeCustomerId } =
+      await stripeService.createCustomerIfNotExist(
+        request.user,
+        userData,
+        userDataService
+      )
+    // If not a new Stripe customer, we want to check for pre-existing bookings
+    if (!newUser) {
+      const activeSession = await stripeService.getActiveSessionForUser(
+        stripeCustomerId,
+        CheckoutTypeValues.BOOKING
+      )
+      if (activeSession) {
+        return {
+          stripeClientSecret: activeSession.client_secret,
+          message: "Existing booking checkout session found"
+        }
+      }
+    }
+
+    const recentActiveSessions = await stripeService.getRecentActiveSessions(
+      CheckoutTypeValues.BOOKING,
+      30
+    )
+
+    const bookingSlotService = new BookingSlotService()
+    const bookingDataService = new BookingDataService()
+
+    const bookingSlots =
+      await bookingSlotService.getBookingSlotsBetweenDateRange(
+        startDate,
+        endDate
+      )
+    // Create a date range list to validate against
+    const daysList = datesToDateRange(
+      firestoreTimestampToDate(startDate),
+      firestoreTimestampToDate(endDate)
+    )
+
+    if (bookingSlots.length !== daysList.length) {
+      this.setStatus(401)
+      return {
+        error: "No booking slot available for one or more dates."
+      }
+    }
+    // iterate through the active stripe sessions and check metadata, reduce the bookingSlots available
+    for (const session of recentActiveSessions) {
+      const bookingSlotMetadata = session.metadata.booking_slots
+      if (bookingSlotMetadata) {
+        for (const bookingSlotId of bookingSlotMetadata) {
+          const bookingSlot = bookingSlots.find(
+            (slot) => slot.id === bookingSlotId
+          )
+          if (bookingSlot) {
+            bookingSlot.max_bookings -= 1
+          }
+        }
+      }
+    }
+    try {
+      Promise.all(
+        daysList.map(async (dateToValidate: Date, index: number) => {
+          // booking slot id and max booking slots
+          const { id, max_bookings } = bookingSlots[index]
+          if (max_bookings < 1) {
+            this.setStatus(401)
+            return {
+              error: "No booking slot available for one or more dates."
+            }
+          }
+          const bookings = await bookingDataService.getBookingsBySlotId(id)
+          if (bookings.some((booking) => booking.user_id === uid)) {
+            this.setStatus(409)
+            return {
+              error: "User already has a booking for this date"
+            }
+          }
+          return dateToValidate
+        })
+      )
```

Comment:
👍, I think there could be a way to write tests for this (just to confirm this logic that only available dates can be booked) - by using the `BookingSlotService` - see the tests written in #382. You could extract this out and write the isolated tests there so we can avoid having to mock the `stripe` calls.

Reply:
Excellent idea! I'll take a look at those.